File: Final/Python/Brisca.py
from os import system
from Mazo import *

class Jugador:
    """
        ! Para leer esto mas legiblemente podemos instalar BetterComments

        * La clase Jugador contiene sus cartas, una funciona que las muestra y getter de las cartas
        * o carta, tambien con una funcion que descarta cierta carta.
    """
    def __init__(self,cartas):
        self.cartasJugador = cartas
    
    def mostrarCartas(self):
        textoCartas = "----> \t" + "\n----> \t".join(map(str, self.cartasJugador))
        print(textoCartas)

    def getCartas(self):
        return self.cartasJugador

    def getCarta(self,i):
        return self.cartasJugador[i]

    def descartarCarta(self,carta):
        self.cartasJugador.remove(carta)


class Brisca:
    """
        ! Para leer esto mas legiblemente podemos instalar BetterComments

        * La clase Brisca maneja todo lo del juego. La creacion del mazo, puntaje
        * el manejo de el turno de un jugador tambien como el mostrar cuales son sus cartas.
        *
        *   ------------------------------------ BRISCA ------------------------------------
        *
        * En el juego de brincan 3 cartas, a cada jugador, y una carta de triunfo debajo de el mazo
        * Dicha carta de triunfo sera la prioridad de Palo.
        * Para jugar los jugadores deben tirar una carta preferentemente del palo de la carta de triunfo
        * Si esta carta es la unica del palo de la carta de triunfo ganara el turno automaticamente
        * En el caso de que dos o mas jugadores pongar el mismo palo de la carta de triunfo se evaluara
        * el valor del mismo, siendo la escala de valores de menor a mayor...
        * Dos -> Cuatro -> Cinco -> Seis -> Siete -> Diez -> Once -> Doce -> Tres -> Uno
        * El jugador con mayor valor ganara el turno
        * En el caso de que ningun jugador juege el mismo palo de la carta de triunfo tambien se evaluara
        * los valores de las cartas.
        * El jugador con mas turnos ganados gana.

        ? El juego representa una mano del juego de Brisca, siendo este 3 turnos.
    """
    def __init__(self):
        print ("A jugar la Brisca!")
        self.mazo = MazoEspanol()
        self.puntaje = 0

    def prepararJuego(self):
        """
            * Esta funcion mezcla el mazo de la Brisca, brinda 3 cartas a 2 jugadores y
            * elige la carta de Triunfo
        """
        print("Empezando mano")
        print("Mezclando mazo...")
        self.mazo.mezclarMazo()
        
        # ! REPARTO DE CARTAS DEL JUGADOR 1
        self.jug1 = Jugador(self.mazo.dar3Cartas())
        print("Cartas del jugador 1:")
        self.jug1.mostrarCartas()
        self.mazo.descartarCartas(self.jug1.getCartas())
        
        # ! REPARTO DE CARTAS DEL JUGADOR 1
        self.jug2 = Jugador(self.mazo.dar3Cartas())
        print("Cartas del jugador 2:")
        self.jug2.mostrarCartas()
        self.mazo.descartarCartas(self.jug2.getCartas())
        
        # ! Extraccion de Triunfo
        self.cartaTriunfo = self.mazo.getTriunfo()
        print("Carta de Triunfo:",self.cartaTriunfo)
        input("Presione ENTER para continuar...")

    def turno(self):
        """
            * Esta funcion es recursiva, el objetivo es que se ejecuta hasta que los jugadores se queden sin cartas
            * Se les mostrara a los jugadores sus cartas, se les dara la opcion de elegir una para jugar.
            * Despues de evaluar las cartas brindadas con el triunfo se decidira un ganador del turno
            * el cual sera notificado y anotado en el sistema de puntaje.
            * Una vez que terminada la mano se mostrara el ganador del juego

            ? Sistema de Puntaje
            ? [-3 -2 -1 0 1 2 3]
            ? Si puntaje > 0 = Gana el jugador 1
            ? Si puntaje < 0 = Gana el jugador 2
            ? Si puntaje == 0 = Es un empate
        """
        system('clear')
        print("-----------------------------------")
        print("Carta de Triunfo:",self.cartaTriunfo)
        print("-----------------------------------")
        print("Turno del jugador 1") 

        self.jug1.mostrarCartas()
        cartaIngreso = int(input("Ingrese el numero de carta que quiere tirar: "))
        cartaJ1 = self.jug1.getCarta(cartaIngreso-1)

        print("La carta elegida es: ", cartaJ1)
        input("Presione ENTER para continuar...")
        print("-----------------------------------")

        print("Turno del jugador 2")   

        self.jug2.mostrarCartas()
        cartaIngreso = int(input("Ingrese el numero de carta que quiere tirar: "))
        cartaJ2 = self.jug2.getCarta(cartaIngreso-1)

        print("La carta elegida es: ", cartaJ2)

        ganador = self.resTurno(cartaJ1,cartaJ2)

        self.jug1.descartarCarta(cartaJ1)
        self.jug2.descartarCarta(cartaJ2)

        self.actualizarPuntaje(ganador)
        self.ganoTurno(ganador)
        
        input("Presione ENTER para continuar...")

        if(len(self.jug1.getCartas()) > 0):
            self.turno()
        else: 
            return 1
    
    def actualizarPuntaje(self,ganador):
        if ganador == 1:
            self.puntaje = self.puntaje + 1
        elif ganador == 2:
            self.puntaje = self.puntaje - 1

    def ganoTurno(self,ganador):
        if ganador > 0:
            print("El turno lo gano el jugador {0}".format(ganador))
        else:
            print("El turno termino en empate")

    def resTurno(self,c1,c2):
        """
            * Esta funcion recibe 2 cartas, siendo de jugadores
            * En esta funcion se decide cual de las dos gana segun las reglas del juego.
            * Siendo prioridad los palos y despues los valores.
        """

        if c1.getPalo() == self.cartaTriunfo.getPalo() and c2.getPalo() != self.cartaTriunfo.getPalo(): return 1
        elif c2.getPalo() == self.cartaTriunfo.getPalo() and c1.getPalo() != self.cartaTriunfo.getPalo(): return 2
        c1Indice = self.mazo.VALORES.index(c1.getValor())
        c2Indice = self.mazo.VALORES.index(c2.getValor())
        if(c1Indice > c2Indice): return 1
        elif(c2Indice > c1Indice): return 2
        else: return 0

    def terminoJuego(self):
        if self.puntaje > 0:
            print("***** Gano el Jugador 1 *****")
        elif self.puntaje < 0:
            print("***** Gano el Jugador 2 *****")
        else:
            print("***** Empate *****")

File: Final/Haskell/Brisca.hs
module Brisca
(   Numero(..),
    Palo(..),
    Carta(..),
    Mazo,
    Jugador,
    paloCarta,
    venceCarta,
    nuevoMazo,
    mezclarMazo,
    empezarJuego,
    descartarCartaMazo,
    descartarCartasJug,
    eliminarItem,
    getTriunfo,
    addPuntaje,
    getPuntaje,
    resTurno,
    mostrarCartas,
    ganoTurno,
    ganoJuego,
    turno) where

import System.Random
import Data.List
import System.Process

data Numero = Dos | Cuatro | Cinco | Seis | Siete | Diez | Once | Doce | Tres | Uno deriving (Read, Enum, Eq, Show, Ord)

data Palo = Copa | Basto | Oro | Espada deriving (Read, Enum, Eq, Show, Ord)

data Carta = Carta Numero Palo deriving (Read, Eq, Show, Ord)

type Mazo = [Carta]
type Jugador = [Carta]

paloCarta :: Carta -> Int
paloCarta (Carta _ Copa) = 1
paloCarta (Carta _ Basto) = 2
paloCarta (Carta _ Oro) = 3
paloCarta (Carta _ Espada) = 4

venceCarta :: Carta -> Carta -> Bool
venceCarta (Carta n1 p1) (Carta n2 p2) = n1 > n2

nuevoMazo :: Mazo
nuevoMazo = [Carta x y | y <- [Copa .. Espada], x <- [Dos .. Uno]]

-- Funcion para mezclar un mazo, recursivo
mezclarMazo :: Mazo -> IO Mazo
mezclarMazo mazo = if not (null mazo)
                       then do
                           let mazoLon = length mazo - 1
                           -- Agarro un numero aleatorio entre 0 y la cantidad de cartas del mazo
                           n <- randomRIO(0,mazoLon) :: IO Int
                           let cartaRandom = mazo !! fromIntegral n
                           mazoMezclado <- mezclarMazo (delete cartaRandom mazo)
                           return (cartaRandom : mazoMezclado)
                       else return mazo

-- Funcion que devuelve una tupla de juego
-- La tupla del juego ( JUGADOR1, JUGADOR2, MAZO, PUNTAJE)
empezarJuego :: Monad m => m Mazo -> m (Jugador, Jugador, Mazo, Int)
empezarJuego m =
    do
    mazo <- m
    return (take 3 mazo, take 3 (drop 3 mazo), drop 6 mazo, 0)

-- Funcion que descarta una carta del mazo tras recibir la tupla del juego
descartarCartaMazo :: [Carta] -> [Carta]
descartarCartaMazo = drop 1

descartarCartasJug :: Jugador-> Int -> Jugador
descartarCartasJug j1 c1 = eliminarItem 0 c1 j1

eliminarItem :: Int -> Int -> [Carta] -> [Carta]
eliminarItem _ _ [] = []
eliminarItem a b (x:xs)
    | a == b = xs
    | otherwise = x : eliminarItem (a+1) b xs

-- Funcion que devuelve una carta, su funcion es devolver la carta de triunfo
getTriunfo :: Mazo -> Carta
getTriunfo = head

addPuntaje :: Int -> Int -> Int
addPuntaje pun ganador
    | ganador == 1 = pun+1
    | ganador == 2 = pun-1
    | otherwise = pun

-- Funcion que devuelve el puntaje segun la tupla del juego
getPuntaje :: (Jugador, Jugador, Mazo, Int) -> Int
getPuntaje (_,_,_,pun) = pun

resTurno :: Carta -> Carta -> Carta -> Int
resTurno c1 c2 triunfo
    | paloCarta c1 == paloCarta triunfo && paloCarta c2 /= paloCarta triunfo = 1
    | paloCarta c2 == paloCarta triunfo && paloCarta c1 /= paloCarta triunfo = 2
    | venceCarta c1 c2 = 1
    | venceCarta c2 c1 = 2
    | otherwise = 0

mostrarCartas :: Jugador -> IO()
mostrarCartas jug1
  | length jug1 == 3 = do
        putStr "Carta 1 -> "
        print(jug1 !! 0)
        putStr "Carta 2 -> "
        print(jug1 !! 1)
        putStr "Carta 3 -> "
        print(jug1 !! 2)
  | length jug1 == 2 = do
   putStr "Carta 1 -> "
   print(jug1 !! 0)
   putStr "Carta 2 -> "
   print(jug1 !! 1)
  | otherwise = do
   putStr "Carta 1 -> "
   print(jug1 !! 0)
    -- putStr "a"

ganoTurno :: Int -> IO()
ganoTurno a
    | a == 0 = putStrLn "Empate!"
    | otherwise = putStrLn ("El jugador " ++ show a ++ " gano el turno!")

ganoJuego :: Int -> IO()
ganoJuego puntaje 
    | puntaje > 0 = putStrLn "\n*** El jugador 1 gano el juego! ***"
    | puntaje < 0 = putStrLn "\n*** El jugador 2 gano el juego! ***"
    | otherwise = putStrLn "\n*** El juego termino en empate! ***"

turno :: (Jugador, Jugador, Mazo, Int) -> IO (Jugador, Jugador, Mazo, Int)
turno (j1,j2,mazo,pun) = do
    
    system "clear"

    putStrLn "============================"
    putStrLn ""
    putStrLn "Cartas del jugador 1: "
    mostrarCartas j1
    
    putStrLn ""
    putStrLn "---------------------------"
    putStrLn ""

    putStrLn "Cartas del jugador 2: "
    mostrarCartas j2
    
    putStrLn ""
    putStrLn "---------------------------"
    putStrLn ""

    let triunfo = getTriunfo mazo
    let mazo' = descartarCartaMazo mazo
    putStr "Carta de Triunfo: "
    print triunfo

    putStrLn ""
    putStrLn "Presione ENTER para continuar..."
    getLine
    system "clear"
    putStrLn "============================"
    putStrLn ""

    putStrLn "Turno del jugador 1: "
    putStrLn ""
    putStr "Carta de Triunfo: "
    print triunfo
    putStrLn ""
    mostrarCartas j1
    putStrLn ""
    putStr "Ingrese el numero de carta que quiere tirar: "
    cartaIngreso <- getLine
    let carta1 = (read cartaIngreso :: Int) - 1
    putStr "La carta elegida es: "
    print(j1 !! carta1)

    putStrLn ""
    putStrLn "Presione ENTER para continuar..."
    getLine
    system "clear"
    putStrLn "============================"
    putStrLn ""

    putStrLn "Turno del jugador 2: "
    putStrLn ""
    putStr "Carta de Triunfo: "
    print triunfo
    putStrLn ""
    mostrarCartas j2
    putStrLn ""

    putStr "Ingrese el numero de carta que quiere tirar: "
    cartaIngreso <- getLine
    let carta2 = (read cartaIngreso :: Int) - 1
    putStr "La carta elegida es: "
    print(j2 !! carta2)

    putStrLn ""
    putStrLn "============================"

    let ganador = resTurno (j1 !! carta1) (j2 !! carta2) triunfo
    let j1' = descartarCartasJug j1 carta1
    let j2' = descartarCartasJug j2 carta2

    let pun' = addPuntaje pun ganador

    putStrLn ""
    ganoTurno ganador

    putStrLn ""
    putStrLn "Presione ENTER para continuar..."
    getLine
    system "clear"
    putStrLn ""

    if not (null j1')
        then turno (j1',j2',mazo',pun')
        else
            return (j1',j2',mazo',pun')
